### PR TITLE
Random seed added as input to samplers under test

### PR DESCRIPTION
**Pull request by yashpote**
If we keep the seed same for each sampling call, the samples are not "independent".
Since only AppMC exposed the seed parameter, I've only used the seed for AppMC.

### CODE REVIEW

File: barbarik.py [hunk 1]
```diff
@@ -171,10 +171,10 @@ def getSolutionFromQuickSampler(inputFile, numSolutions, indVarList):
 
     @staticmethod
     def getSolutionFromUniform(inputFile, numSolutions):
-        return SolutionRetriver.getSolutionFromMUSE(inputFile, numSolutions)
+        return SolutionRetriver.getSolutionFromSPUR(inputFile, numSolutions)
```
Comment (msoos) [inline barbarik.py]: Wait, what's going on here? Was this a bug? Also, this should be a separate change. This has nothing to do with seeds.

File: barbarik.py [hunk 2]
```diff
@@ -171,10 +171,10 @@ def getSolutionFromQuickSampler(inputFile, numSolutions, indVarList):
...
 
     @staticmethod
-    def getSolutionFromMUSE(inputFile, numSolutions):
+    def getSolutionFromSPUR(inputFile, numSolutions):
```
Comment (msoos) [inline barbarik.py]: Wait, what's going on here? Was this a bug? Also, this should be a separate change. This has nothing to do with seeds.

File: barbarik.py [hunk 3]
```diff
@@ -577,9 +577,12 @@ def one_experiment(self, experiment, j, i):
         if self.thresholdSolutions < self.minSamples:
             return None, None
 
+        # generate a new seed value for every call to the sampler
+        preSeed = random.randint(1,j*1000)
+        newSeed = random.randint(preSeed, i*j*1000)
```
Comment (msoos) [inline barbarik.py]: What's this? You don't need this at all. Just `newSeed = random.randint(1, X)`. What's the point of j here? What's the point of preSeed? None of this makes sense.

File: barbarik.py (hunk 1 above)
Comment (yashpote) [reply barbarik.py]: We are using SPUR as the uniform sampler, not MUSE.

File: barbarik.py (hunk 2 above)
Comment (yashpote) [reply barbarik.py]: We are using SPUR as the uniform sampler, not MUSE. Yes, I will make a separate pull request.

**Comment by msoos** [comment]
Thanks! Also, please use the seeds. And I'm still not sure we need the
seeds. So there is a lot of confusion here. Please clean it all up :)
Thanks!

**Comment by msoos** [comment]
Hi,

Thanks for that fix! It's cleaner now :) However, it still has some issues:
- It does not use the seed that is passed in.
- As per my note on the change, I am unsure that the complicated preseed/seed setup is needed -- a single random number generation, creating a seed to be passed to the generators is I believe enough.

Finally, I am not convinced we need to use seeds. We only do a measurement on a the same CNF only once -- so we will never measure the same CNF twice. The sampler therefore will get different CNFs, and will use the same internal, fixed, seed. Hence, it should be OK, no? Have you asked Kuldeep about this? I am not sure we need to pass different seeds. This is kind of a theoretical question, I think, so I'm not the authority, so I'm unsure what's the right thing to do.

File: barbarik.py (hunk 1 above)
Comment (msoos) [reply barbarik.py]: Oh, I see! Can you make this into a separate pull request, I'd love to merge this in!

**Comment by msoos** [comment]
Hi Yash,

Have you worked on this lately? I just need the DIFF to:
* actually make use of the seeds -- you have put all this work into adding them, let's use them!
* simplifying the seed generation. It doesn't need 2 rounds of rand(), one rand() call should do the trick :)

Looking forward to the above, minor, changes so I can merge this request in! Thanks :)

**Comment by yashpote** [comment]
> * It does not use the seed that is passed in.

Ideally, we want the samplers to expose the seed for reproducibility. However, only AppMC3 exposes the seed. So we need to pass the seed to AppMC3 and the custom sampler by default.

I have updated the code to only send the seed to the samplers which use it.   

> * As per my note on the change, I am unsure that the complicated preseed/seed setup is needed -- a single random number generation, creating a seed to be passed to the generators is I believe enough.

For AppMC3, we frequently run the experiments in parallel. If we keep the seed same, every call to AppMC3 would sample identically on the formula hence violating the independence assumption. 
However, this code is clearly sequential so I'm not sure about whether it's right to implement the seed thing here.

Also, you are right, there was a one-line solution for the random seed. I have updated that in the latest commit.
 
> Finally, I am not convinced we need to use seeds. We only do a measurement on the same CNF only once -- so we will never measure the same CNF twice. The sampler, therefore, will get different CNFs, and will use the same internal, fixed, seed. Hence, it should be OK, no? Have you asked Kuldeep about this? I am not sure we need to pass different seeds. This is kind of a theoretical question, I think, so I'm not the authority, so I'm unsure what's the right thing to do.

The formula depends on the assignments sampled from the CNF. It is not impossible for the same assignment to get sampled more than once in the same sampling run by the same sampler. So it is possible that the same CNF gets generated more than once in one run of Barbarik. 

We need Barbarik to run such that each call to the sampler is independently random, thus I have used a seed that depends on all of the three loop iterations.

**Comment by msoos** [comment]
I see! Thanks for the fixes and the explanation :) Merging!

Thanks agin!

**Pull request merged**